src/devices/bus/msx/cart/slotexpander.cpp

```diff
@@ -0,0 +1,152 @@
+// license:BSD-3-Clause
+// copyright-holders:Wilbert Pol
+#include "emu.h"
+#include "slotexpander.h"
+#include "machine/input_merger.h"
+
+/*
+Emulation of a 4 slot expander for the MSX system.
+
+Some slot expanders have the ability to enable/disable subslots through
+jumpers/switches, or adjustable volume per slot. None of that is emulated.
+
+Known slot expanders:
+- 8bits4ever Slot x4 - audio mixer with adjustable volume per slot not emulated.
+- CIEL Mini Slot Expander - Jumpers and switches not emulated.
+- Digital Design Expansor de slots DDX
+- ECC Expansion Computer Case - 8 slot version not emulated
+- Front Line Slot Expander
+- G.DOS Slotexpander
+- Hans Oranje slotexpander - Switches not emulated
+- Incompel Expansor de slots
+- MAD Expander Slot Box MXE-MAIN-A / MXE-MAIN4-A - Switches not emulated
+- MK Slotexpander - Switches not emulated
+- MSX Club Gouda Slotexpander - Switches not emulated
+- Mitsubishi ML-20EB
+- Neos EX-4
+- Padial LPE-4EXP-V3SC
+- Padial LPE-4EXP-V4SC
+- Repro Factory My Super Expander 4X
+- Sinfox Slotexpander
+- Sony HBI-50 - Only expands to 2 slots
+- Sunrise 8x SlotExpander - only 4 slots emulated
+- Supersoniqs Modulon
+- Toshiba HX-E601
+- Victor HC-A703E - Only expands to 2 slots?
+- Zemina Expansion slot - Only expands to 2 slots
+
+*/
+
+namespace {
+
+class msx_cart_slotexpander_device : public device_t, public msx_cart_interface
+{
+public:
+	msx_cart_slotexpander_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+		: device_t(mconfig, MSX_CART_SLOTEXPANDER, tag, owner, clock)
+		, msx_cart_interface(mconfig, *this)
+		, m_cartslot(*this, "cartslot%u", 0)
+		, m_irq_out(*this, "irq_out")
+		, m_view0(*this, "view0")
+		, m_view1(*this, "view1")
+		, m_view2(*this, "view2")
+		, m_view3(*this, "view3")
+		, m_secondary_slot(0)
+	{ }
+
+protected:
+	virtual void device_start() override;
+	virtual void device_add_mconfig(machine_config &config) override;
+	virtual void device_resolve_objects() override;
+
+private:
+	template<int Slot> void add_cartslot(machine_config &mconfig);
+	u8 secondary_slot_r();
+	void secondary_slot_w(u8 data);
+
+	required_device_array<msx_slot_cartridge_device, 4> m_cartslot;
+	required_device<input_merger_any_high_device> m_irq_out;
+	memory_view m_view0;
+	memory_view m_view1;
+	memory_view m_view2;
+	memory_view m_view3;
+	u8 m_secondary_slot;
+};
+
+template<int Slot>
+void  msx_cart_slotexpander_device::add_cartslot(machine_config &mconfig)
+{
+	MSX_SLOT_CARTRIDGE(mconfig, m_cartslot[Slot], 0);
+	// During validate_device_types the finders from the parent cartridge slot are not present.
+	if (required_maincpu())
+	{
+		m_cartslot[Slot]->set_memory_space(*required_maincpu(), AS_PROGRAM);
+		m_cartslot[Slot]->set_io_space(*required_maincpu(), AS_IO);
+		m_cartslot[Slot]->set_maincpu(*required_maincpu());
+	}
```

Comment:
It’s generally not a great idea to do this at the point of adding a device because object finders can be retargeted.  You probably want to do it in `device_config_complete`.

I also think you’re going to have issues because the cartridge devices assume they can install taps in these spaces to respond to CPU memory and I/O accesses, but those taps have no awareness of whether a particular subslot is selected, so they’ll respond to accesses when they shouldn’t.

Reply:
If I move this to device_config_complete then I always run into this assert:
Assertion failed: (m_target), function operator->, file devfind.h, line 478.
when attempting to set any of the address spaces or maincpu on the new cartridge devices.

Reply:
> I also think you’re going to have issues because the cartridge devices assume they can install taps in these spaces to respond to CPU memory and I/O accesses, but those taps have no awareness of whether a particular subslot is selected, so they’ll respond to accesses when they shouldn’t.

The memory space is only provided to accommodate the banking in superloderunner. That one banks on writes to address 0000 regardless of which slot or subslot has been selected.

In general MSX cartridges should install their memory accesses in the views provided by page(0), page(1), etc in the msx_cart_interface. Then they will respond correctly based on the selections of slot and subslot.

For I/O accesses most devices will always respond regardless of whether a slot has been selected.

Reply:
> If I move this to device_config_complete then I always run into this assert:
Assertion failed: (m_target), function operator->, file devfind.h, line 478.
when attempting to set any of the address spaces or maincpu on the new cartridge devices.

You shouldn’t be dereferencing an object finder at that point – you want the other object finder to take on its base device and tag directly.  What’s the actual call stack to the failed assertion?  Which object finder is being dereferenced?

> For I/O accesses most devices will always respond regardless of whether a slot has been selected.

Do the slot expanders really not gate I/O accesses depending on the selected slot?  Wouldn’t that make collisions likely and cause many combinations of cartridges to break?

Reply:
> > For I/O accesses most devices will always respond regardless of whether a slot has been selected.
> 
> Do the slot expanders really not gate I/O accesses depending on the selected slot? Wouldn’t that make collisions likely and cause many combinations of cartridges to break?

No, they don't. Also multiple 16K pages from the different subslot can be active at the same time in the Z80's address space. That is also why memory mapped I/O is preferred in these systems since that only hardware from active pages can be accessed by the Z80.
This is not specific to slotexpanders, the same behaviour occurs between cartridges inserted in main cartridge slots.

Reply:
> > If I move this to device_config_complete then I always run into this assert:
> > Assertion failed: (m_target), function operator->, file devfind.h, line 478.
> > when attempting to set any of the address spaces or maincpu on the new cartridge devices.
> 
> You shouldn’t be dereferencing an object finder at that point – you want the other object finder to take on its base device and tag directly. What’s the actual call stack to the failed assertion? Which object finder is being dereferenced?
> 

The assert triggers on m_cartslot[0-3]. Which in device_config_complete appears to not be resolved even though it has just been set up during device_add_mconfig?
I must be doing something very wrong somewhere...

I put this in device_config_complete:
```
void msx_cart_slotexpander_device::device_config_complete()
{
	printf("device_config_complete\n");
	if (get_cpu_finder() && get_cpu_finder()->target())
	{
		for (int subslot = 0; subslot < 4; subslot++)
		{
			m_cartslot[subslot]->set_memory_space(*get_cpu_finder(), AS_PROGRAM);
			m_cartslot[subslot]->set_io_space(*get_cpu_finder(), AS_IO);
			m_cartslot[subslot]->set_maincpu(*get_cpu_finder());
		}
	}
}
```

and this is the trace I get
```
device_config_complete
Assertion failed: (m_target), function operator->, file devfind.h, line 478.
Process 68808 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #4: 0x00000001013fca0c msxd`object_finder_common_base<msx_slot_cartridge_device, true>::operator->(this=0x000000011a008b18) const at devfind.h:478:36
   475 		/// Allows pointer-member-style access to members of the target
   476 		/// object.  Asserts that the target object has been found.
   477 		/// \return Pointer to target object if found, or nullptr otherwise.
-> 478 		ObjectClass *operator->() const { assert(m_target); return m_target; }
   479 	
   480 	protected:
   481 		/// \brief Designated constructor
Target 0: (msxd) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #0: 0x00000001994f2868 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x0000000199529cec libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x00000001994622c8 libsystem_c.dylib`abort + 180
    frame #3: 0x0000000199461620 libsystem_c.dylib`__assert_rtn + 272
  * frame #4: 0x00000001013fca0c msxd`object_finder_common_base<msx_slot_cartridge_device, true>::operator->(this=0x000000011a008b18) const at devfind.h:478:36
    frame #5: 0x00000001013facd4 msxd`(anonymous namespace)::msx_cart_slotexpander_device::device_config_complete(this=0x000000011a008800) at slotexpander.cpp:103:4
    frame #6: 0x0000000101dcf8b8 msxd`device_t::config_complete(this=0x000000011a008800) at device.cpp:310:2
    frame #7: 0x0000000104053684 msxd`validity_checker::validate_devices(this=0x000000016fdfe4a0, config=0x000000016fdf90c8) at validity.cpp:2623:15
    frame #8: 0x00000001040495cc msxd`validity_checker::validate_one(this=0x000000016fdfe4a0, driver=0x0000000105425cd8) at validity.cpp:1956:3
    frame #9: 0x00000001040498dc msxd`validity_checker::check_shared_source(this=0x000000016fdfe4a0, driver=0x0000000105426a40) at validity.cpp:1822:4
    frame #10: 0x000000010105e068 msxd`mame_machine_manager::execute(this=0x0000600002c04580) at mame.cpp:275:10
    frame #11: 0x00000001000f53ac msxd`cli_frontend::start_execution(this=0x000000016fdfef40, manager=0x0000600002c04580, args=size=8) at clifront.cpp:277:22
    frame #12: 0x00000001000f6438 msxd`cli_frontend::execute(this=0x000000016fdfef40, args=size=8) at clifront.cpp:293:3
    frame #13: 0x000000010105fbd0 msxd`emulator_info::start_frontend(options=0x000000016fdff0e0, osd=0x000000016fdfef90, args=size=8) at mame.cpp:454:18
    frame #14: 0x000000010427b388 msxd`main(argc=8, argv=0x000000016fdff588) at sdlmain.cpp:100:9
    frame #15: 0x00000001991ffe50 dyld`start + 2544
(lldb) up
frame #5: 0x00000001013facd4 msxd`(anonymous namespace)::msx_cart_slotexpander_device::device_config_complete(this=0x000000011a008800) at slotexpander.cpp:103:4
   100 		{
   101 			for (int subslot = 0; subslot < 4; subslot++)
   102 			{
-> 103 				m_cartslot[subslot]->set_memory_space(*get_cpu_finder(), AS_PROGRAM);
   104 				m_cartslot[subslot]->set_io_space(*get_cpu_finder(), AS_IO);
   105 				m_cartslot[subslot]->set_maincpu(*get_cpu_finder());
   106 			}
(lldb) print subslot
(int) $0 = 0
(lldb) print m_cartslot[0]
(device_finder<msx_slot_cartridge_device, true>) $1 = {
  object_finder_base<msx_slot_cartridge_device, true> = {
    object_finder_common_base<msx_slot_cartridge_device, true> = {
      finder_base = {
        m_next = nullptr
        m_base = {
          __f_ = 0x000000011a008800
        }
        m_tag = 0x000000011a008ab8 "cartslot0"
        m_resolved = false
      }
      m_target = nullptr
    }
  }
}
```

Reply:
OK, there are two issues here:
* You can’t expect `->target()` to work at that point because object finders have not been resolved.
* You can’t dereference `m_cartslot` at that point because object finders have not been resolved – you need to use `subdevice<...>(...)`.

Reply:
Ah, I see, that makes sense. I was just confused by the fact that the finders did work during device_add_config and didn't later on during device_config_complete.
Thanks.